#1090 Creation date added to Dy/Xe/RtAttachment and tests

CarlosGines opened the pull request:
Implementations for #1090:
- Added Attachment creation date to DynamoDB.
- Added Attachment creation date to Xembly generated files.
- Enabled integration tests for attachment creation date.

karato commented:
@carlosgines Thanks for your pull request, let me find someone who can review it

CarlosGines commented:
@karato please find someone to review this PR

karato commented:
@original-brownbear do this review

original-brownbear commented:
@CarlosGines I'm on it.

original-brownbear commented on netbout-web/src/main/java/com/netbout/dynamo/DyAttachments.java:
```diff
@@ -98,6 +98,11 @@
     /**
+     * Creation time of the attachment.
+     */
+    public static final String ATTR_DATE = "date";
```
@CarlosGines please don't add new public static literals, see [here](http://www.yegor256.com/2015/07/06/public-static-literals.html).

original-brownbear commented on netbout-web/src/main/java/com/netbout/dynamo/DyAttachment.java:
```diff
@@ -163,18 +163,11 @@ public boolean unseen() throws IOException {
         return new Date(
             Long.parseLong(
-                this.item.get(DyFriends.ATTR_UPDATED).getN()
+                this.item.get(DyAttachments.ATTR_DATE).getN()
```
@CarlosGines please don't add new public static literals, see [here](http://www.yegor256.com/2015/07/06/public-static-literals.html).

original-brownbear commented on netbout-web/src/main/java/com/netbout/dynamo/DyAttachments.java:
```diff
@@ -236,6 +241,7 @@ public void create(final String name) throws IOException {
                 .with(DyAttachments.ATTR_CTYPE, "text/plain")
                 .with(DyAttachments.ATTR_DATA, " ")
                 .with(DyAttachments.ATTR_ETAG, "empty")
+                .with(DyAttachments.ATTR_DATE, System.currentTimeMillis())
```
@CarlosGines please inline `"date"` here, remove the public static literal and handle the test accordingly. See [here](http://www.yegor256.com/2015/07/06/public-static-literals.html) for suggestions on how to do it right.

original-brownbear commented on netbout-client/src/main/java/com/netbout/client/RtAttachment.java:
```diff
@@ -119,7 +119,17 @@ public boolean unseen() throws IOException {
     @Override
     public Date date() throws IOException {
-        return new Date();
+        return new Date(
```
@CarlosGines I suggest adding  a puzzle here to remove the duplication of this section here:
```java
         this.request.fetch()
            .as(RestResponse.class)
            .assertStatus(HttpURLConnection.HTTP_OK)
            .as(XmlResponse.class)
            .xml()
            .xpath(this.xpath("ctype/text()"))
            .get(0);
```

It's ok that this got copy and pasted here, in fact anything but doing so would be wrong according to our principles, but we should ensure that this get's cleaned up down the line imo :)

CarlosGines replied:
@original-brownbear I am not sure of what you mean here, do you mean that there is duplication in my code regrading `RtAttachment.ctype()`? Or the same duplications as in `RtAttachment.etag()`, `RtAttachment.unseen()`, `RtAttachment.author()`? If it is not specific to my code, you can create a bug about it, I wouldn't consider it specific to this task, nor a puzzle, since there is nothing incomplete in the code in here.

original-brownbear replied:
@CarlosGines I agree that his is not specific to the task, I even tried to point this out :)

You know I just think it's more in line with the spirit of how we work to put a puzzle here because:
1. We have some problem with duplication (not a big one, but still)
2. This PR makes it slightly worse ( again not your fault, fix is fine here ), so this PR certainly would need to have that duplication cleaned up to be "perfectly fine".
3. Hence the puzzle would make this PR "perfect" if solved.

... believe me I would create the bug and rake in the 15 min if I thought it was appropriate ;)

original-brownbear commented:
@CarlosGines done, please see above remarks.

original-brownbear commented:
@CarlosGines see my note on the duplication. I think the very same applies to the public static literals, puzzle is fine imo :)

CarlosGines commented:
@original-brownbear here is a new commit regarding the public static literal. Although there is an unrelated error on the CI build, tell me do you consider this better?

original-brownbear commented:
@CarlosGines well don't ask it like that with the "better", you saw the link ;) I think it's more in line with what is put forward in that article, which I tend to agree with in general. This is a bit of a special case since there is already a few public static literals here. As I said puzzle and all literals or using it like this would both be acceptable solutions imo.

CarlosGines commented:
@original-brownbear new modifications:
- Public static literal is back.
- Added puzzle about public static literals in `DyAttachments`
- Added puzzle about code duplication in `RtAttachment`

Earlier I asked with the "better" because I really want to know your opinion. I do have read the article, but I still consider having a public static literal better than having a hardcoded string in several places. I would be happy to see a solution that is non of those, but the article suggest creating an object, which I don't see fit here either. So, puzzle created about it.

Thank you for your messages.

original-brownbear commented on netbout-client/src/main/java/com/netbout/client/RtAttachment.java:
```diff
@@ -50,6 +50,9 @@
  * @since 2.0
+ * @todo #1090:30min remove code duplication in public methods in this class
+ *  except in method name(). Create one or more private methods to achieve
+ *  this.
```
@CarlosGines please rephrase this a little :)
Give the explicit names of those methods containing duplication and remove the second sentence, it kind of says nothing if you think about it :)

original-brownbear commented on netbout-web/src/main/java/com/netbout/dynamo/DyAttachments.java:
```diff
@@ -46,10 +46,11 @@
  * @version $Id$
  * @since 2.0
+ * @todo #1090:30min No public static literals are allowed according to our
+ *  methodology. This class should be refactored to not contain any of them.
```
@CarlosGines I think we should also rephrase this a little, saying "public static literals are not recommended" and I would also recommend adding the link to http://www.yegor256.com/2015/07/06/public-static-literals.html. Otherwise this might be very confusing :)

original-brownbear commented:
@CarlosGines looks good, just two suggestions for rephrasing the puzzles, always try to imagine you're the performer of those ;) Always good to have very clear instructions including names of functions on **what** is to be fixed, but stay flexible on the **how** :)

CarlosGines commented:
@original-brownbear so you want me to modify the puzzle in `RtAttachment` I am guessing, it is the only one where I am telling how to solve it. Please be more specific on what you want to be fixed for the code to be accepted.

CarlosGines commented:
@original-brownbear please see new commit modifying `RtAttachment` puzzle description.

original-brownbear commented:
@CarlosGines please just also relax the language as instructed in my comment on the other puzzle and we're fine :)

CarlosGines commented:
@original-brownbear I am sorry, I did not see your specific comments on puzzles before, I don't know why. New commit, this should be it!

original-brownbear commented:
@CarlosGines no worries :) Looks good to me now :)

original-brownbear commented:
@rultor merge please :)

rultor commented:
> @rultor merge please :)

@original-brownbear Thanks for your request. @dmzaytsev Please confirm this.

original-brownbear commented:
@dmzaytsev ping :)

dmzaytsev commented:
@rultor merge

rultor commented:
> @rultor merge

@dmzaytsev OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/7976-205960619)

rultor commented:
> @rultor merge

@dmzaytsev Done! FYI, the full log is [here](http://www.rultor.com/t/7976-205960619) (took me 7min)

original-brownbear commented:
@rultor deploy please :)

rultor commented:
> @rultor deploy please :)

@original-brownbear Thanks for your request. @dmzaytsev Please confirm this.

dmzaytsev commented:
@rultor deploy

rultor commented:
> @rultor deploy

@dmzaytsev OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/7976-206067901)

rultor commented:
> @rultor deploy

@dmzaytsev Done! FYI, the full log is [here](http://www.rultor.com/t/7976-206067901) (took me 8min)

karato commented:
@ypshenychka please, review this issue for QA compliance, as per [par.24](http://at.teamed.io/policy.html#24)

ypshenychka commented:
@karato Quality is good.

karato commented:
> @karato Quality is good.

@ypshenychka thanks a lot

karato commented:
@original-brownbear quality is good here, 10 mins added to @dmzaytsev (our architect) in transaction 82799704

original-brownbear commented:
@karato what about me ? :)

karato commented:
@original-brownbear added 10 mins to @ypshenychka, for QA review, payment code is `82803454`; Thanks so much! Your account was topped up for **33 mins** (transaction ID is `AP-4G43279819748745R`, the task took 104 hours and 53 mins); you're getting extra minutes here (c=18); added +33 to your rating, now it is equal to [+2887](http://www.netbout.com/b/36646?open=rating)

karato commented:
@rultor deploy now pls

rultor commented:
> @rultor deploy now pls

@karato OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/7976-213477598)

rultor commented:
> @rultor deploy now pls

@karato Done! FYI, the full log is [here](http://www.rultor.com/t/7976-213477598) (took me 17min)
